### ISSUE TITLE
get `kselftests` target building

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1698]
Looking at #906 , I noticed that `make LLVM=1 -j72 kselftest` fails with tons of warnings and errors; maybe we can use this as a metabug, since there will probably be lots of fixes necessary for this.

**Comment by nickdesaulniers** [comment]
cc @gctucker , looking at #374 I thought we had kselftests building with clang...

**Comment by gctucker** [comment]
Well there are a bunch of errors but some tests are building OK as far as I can tell:
https://storage.staging.kernelci.org/kernelci/staging-next/staging-next-20220830.0/x86_64/x86_64_defconfig+kselftest/clang-16/logs/kselftest.log

**Comment by gctucker** [comment]
Note that there are 2 fixes applied on this KernelCI staging-next branch, for when the arch is `x86_64` or `i386` rather than just the ambiguous `x86`:
https://github.com/kernelci/linux/commit/839f7af56f19414429e69c0c4b06eb76ad985e84
https://github.com/kernelci/linux/commit/bbda7f967df09b1c3f87866e75169ccd801596b6

They've been reviewed on the mailing list by @nathanchance so in principle it's just a matter of time until they get applied to the kselftest tree.  Not sure if that's related to the issues you're seeing, but worth mentioning anyway.

**Comment by JustinStitt** [comment]
Here's a tentative series from the mailing list that Benjamin and I have worked on that solves (at least) my kselftest build issues:

https://lore.kernel.org/all/20230825182316.m2ksjoxe4s7dsapn@google.com/

After discussion, Benjamin proposed a 3-patch series which fixed _some_ of the issues I was experiencing. I, then, added a 4th patch which should be applied on top of this series. Hopefully Benjamin can re-roll this into an n=4 patch series so that `b4 shazam` applies it cleanly.

Once this is re-rolled into a single n=4 series (and subsequently applied to next) I think we can close this issue 😸

**Comment by JustinStitt** [comment]
Update: I pinged Benjamin off-list regarding the state of this series. Fingers crossed 🤞

**Comment by JustinStitt** [comment]
Just to keep this issue up-to-date...

Here are some patches in the works:
* [[PATCH v2 0/3] selftests/hid: fix building for older kernels](https://lore.kernel.org/r/20230908-kselftest-09-08-v2-0-0def978a4c1b@google.com)

* [[PATCH RFC] selftests/rseq: fix kselftest Clang build warnings](https://lore.kernel.org/r/20230908-kselftest-param_test-c-v1-1-e35bd9052d61@google.com)

**Comment by JustinStitt** [comment]
Another update, Benjamin has the hid/bpf fixes applied: https://lore.kernel.org/all/169703194378.3126081.13251971976785038026.b4-ty@kernel.org/

**Comment by JustinStitt** [comment]
The plot thickens... It looked like we were finally going to get some clean builds with an incantation like: 

`$ make LLVM=1 -j128 headers && make LLVM=1 -j128 -C tools/testing/selftests`

... but 

Commit 14571ab1ad213 ("kselftest: Add new test for detecting unprobed Devicetree devices")

which is in v6.7-rc1 has introduced some problems.

For me, the build just hangs when it gets to the `dt` selftest. For others, it seems some errors are produced: https://lore.kernel.org/all/e90cb52f-d55b-d3ba-3933-6cc7b43fcfbc@arm.com/ regarding infinite symlinks.

**Comment by gctucker** [comment]
> Commit [14571ab](https://github.com/ClangBuiltLinux/linux/commit/14571ab1ad213de59b3726a40aea7ca0365bf445) ("kselftest: Add new test for detecting unprobed Devicetree devices")
> 
> which is in v6.7-rc1 has introduced some problems.

Yes I saw it got stuck in an infinite loop in the Python script.

@nfraprado Looks like it wasn't just me seeing this problem, do you need some help to fix it or reproduce it?

**Comment by nfraprado** [comment]
@gctucker Hi, I just got back from vacation and saw that report in the list regarding the infinite symlinks. I'll look into it and report back.

**Comment by nfraprado** [comment]
I've replied to the thread on the mailing list, but just to close the loop, I've sent this patch fixing the issue:
https://lore.kernel.org/all/20231107225624.9811-1-nfraprado@collabora.com/

**Comment by johnhubbard** [comment]
Hi everyone, @nickdesaulniers asked me to respond here. I recently posted 30 or 35 patches to fix all clang errors and warnings in the kselftests. Status:

11 patches are already merged into Linux main:
commit 14d28ec6f821 (selftests/resctrl: fix clang build warnings related to abs(), labs() calls)
commit eb709b5f6536 (selftests/net: fix uninitialized variables)
commit 41b307ad756e (bpftool, selftests/hid/bpf: Fix 29 clang warnings)
commit d8171aa4ca72 (selftests/resctrl: fix clang build failure: use LOCAL_HDRS)
commit 019baf635eb6 (selftests/binderfs: use the Makefile's rules, not Make's implicit rules)
commit 8f6d24a5db2a (selftests/cgroup: fix uninitialized variables in test_zswap.c)
commit 3309ca6f47f1 (selftests/cgroup: cpu_hogger init: use {} instead of {NULL})
commit 0515089418d0 (selftests/cgroup: fix clang warnings: uninitialized fd variable)
commit 1da2363228d6 (selftests/cgroup: fix clang build failures for abs() calls)
commit 580ea358af0a (selftests/mm: fix additional build errors for selftests)
commit e076eaca5906 (selftests: break the dependency upon local header files)

These patches are pending. Most have been reviewed, with the notable exception of the x86 subsystem. I'll rebase on top of 6.10-rc1 and repost at that time (Monday):

[PATCH 0/7] selftests/x86: fix build errors and warnings found via clang [1]
[PATCH] selftests/vDSO: fix clang build errors and warnings
[PATCH] selftest/timerns: fix clang build failures for abs() calls
[PATCH] selftests/timers: remove unused irqcount variable
[PATCH] selftests/user_events: silence a clang warning: address of packed member
[PATCH] selftests/alsa: make dump_config_tree() as void function
[PATCH v2] selftests/netfilter: return a value for several "int" functions
[PATCH v2] selftests/mqueue: fix 5 warnings about signed/unsigned mismatches
[PATCH 1/2] selftests/openat2: fix clang build failures: -static-libasan, LOCAL_HDRS
[PATCH 2/2] selftests/fchmodat2: fix clang build failure due to -static-libasan
[PATCH 0/3] selftests/futex: clang-inspired fixes
[PATCH 1/2] selftests/net: suppress clang's "variable-sized type not at the end" warning

Here's a few more that are most likely resolved by other authors' patches:

[PATCH v2] selftests/alsa: missing a return value in unused dump_config_tree()
[PATCH] selftests/exec: build with -fPIE instead of -pie, to make clang happy
Re: [PATCH v2] selftests: exec: make binaries position independent
[PATCH 0/3] binfmt_elf: Honor PT_LOAD alignment for static PIE

[1] https://lore.kernel.org/20240503030214.86681-1-jhubbard@nvidia.com
...I can post the other links if requested, but everything is Cc'd to linux-selftests already.
